### PR TITLE
Add mach bootstrap-android and test-android-startup commands

### PR DESCRIPTION
**Pull request by SimonSapin**
This adds a `./mach boostrap-android` subcommand that downloads and installs the tools, SDK, NDK,  emulator, and system image for Android. In an environment that can build Servo at all, this should be enough to get all additional dependencies to cross-compile to Android, package an APK, and load it onto a device or an emulator.

At the moment it requires an interactive user to accept the license (and confirm no customization of the emulated virtual device hardware), and then prints environment variables to set for `mach` as well as the command to run to start the emulator (with an already-configured image). A possible next step could be to automate all this, and have `./mach build` run it automatically when needed. (I don’t know if auto-accepting the license is something we should do though.)

This also adds `--emulator` and `--usb` parameters to `./mach install --android` and `./mach run --android`, which tell `adb` what device to pick when both are present. And makes `./mach run --android` print the new process’s PID, for use with e.g. `adb -e logcat --pid 2263`.

Finally, adds the `./mach boostrap-android` subcommand which starts an emulator, installs the APK (it assumes that `mach build` and `mach package` were already executed), runs a single HTML test case, and checks for a message coming from JS through `console.log()` and `adb logcat`.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/21094)

### CODE REVIEW

**Comment by SimonSapin** [comment]
r? @jdm

With this I can compile Servo and run it on an hardware phone and have web content show up on the screen. In the emulator however, I get a crash at startup and logcat contains `RustAndroidGlueStdouterr: Failed to create window.: OsError("eglChooseConfig failed")` (from glutin).

This means that none of the available EGL configurations satisfy what is requested by Glutin, but there is no error message to say more about this. The stand-alone program shows that answer is most likely the 0x40 bit of `EGL_RENDERABLE_TYPE`, which indicates support for OpenGL ES 3. It is requested by Glutin, and present in many of the 64 configs available on my Nexus 5x but none of the 4 configs on this emulator.

I am starting the emulator with [`-gpu swiftshader_indirect`](https://developer.android.com/studio/run/emulator-acceleration#command-gpu), which enables a software implementation of GL. (`-gpu host` also exists, but likely won’t work AWS servers without a hardware GPU and as far as I know only exists for Android x86.)

[Release notes](https://developer.android.com/studio/releases/emulator#27-1-8) claim that the emulator gained a few versions ago support for GLES3 in Swiftshader, but ["Extended controls" docs](https://developer.android.com/studio/run/emulator#searchbox) describes a configurable choice:

> OpenGL ES API level: Select the maximum version of OpenGL ES to use in the emulator.
>
> * Autoselect: Let the emulator choose the OpenGL ES version based on the host and guest support.
> * Renderer maxium (up to OpenGL ES 3.1): Attempt to use the maximum version of OpenGL ES.
> * Compatibility (OpenGL ES 1.1/2.0): Use the version of OpenGL ES that is compatible with most environments.

Unfortunately this seems to be a GUI-only config, there is no command-line equivalent documented there.

I feel like I’m running out leads on the software-GL-in-emulator front. In the meantime I can run Servo on a hardware device to look at plumbing wptrunner and adb-logcat together.

**Comment by jdm** [comment]
I recommend using @paul as a resource, since I'm going to be returning to stealth vacation mode. However, are you able to run Servo in an emulator? I certainly was able to as part of my android-x86 work (on Mac), and it worked (slowly) for ARM builds too. One important change in my setup was using the most recent system image for the AVD device, rather than one that matched the API level that we build with.

**Comment by SimonSapin** [comment]
No, `eglChooseConfig` doesn’t succeed in the emulator so I get a startup crash. That’s good to know re system image versions, I’ll try it. It’s likely that the emulator on mac was using your hardware GPU so we’d need to make a different configuration work, for CI. (Unless we want to run those jobs on macs, since they presumably have a GPU?)

r? @paul

**Comment by SimonSapin** [comment]
No luck either with `system-images;android-28;google_apis;x86`, `-gpu host` and running under Xorg (which should be the "normal Linux desktop" case for accessing the hardware GPU).

**Comment by SimonSapin** [comment]
Ah I see:

https://android-developers.googleblog.com/2017/05/android-studio-3-0-canary1.html

> To use OpenGL ES 3.0 with the Android Emulator, your development machine needs a host GPU graphics card that supports OpenGL 3.2 or higher on Microsoft® Windows® or Linux (with Apple MacOS® support coming in the future). 

But `glxinfo|grep 'OpenGL version'` on my dev machine is `OpenGL version string: 3.0 Mesa 18.0.0-rc5`

**Comment by SimonSapin** [comment]
Success! When starting the emulator with `-gpu guest` (which is undocumented‽) and with this glutin patch https://github.com/tomaka/glutin/pull/1036, I can get Servo to render web content in an emulator. It seems to even be working without an X11 server (and so without a hardware GPU), when also passing `-no-window`.

**Comment by SimonSapin** [comment]
However this is with x86. Still no GLES3 on emulated ARM.

**Comment by paulrouget** [comment]
I'll be looking at this soon.

About the emulator, I never managed to get servo to work with an Arm emulator. As for x86, it works here both on Mac and linux.

Also, you copy pasted a type multiple time: s/intall/install/

**Review by paulrouget — COMMENTED** [review]
I get this message:

```
Parsing /home/Paul/git/servo/android-toolchains/sdk-tools-linux-4333796/build-tools/25.0.2/package.xmlParsing /home/Paul/git/servo/android-toolchains/sdk-tools-linux-4333796/emulator/package.xmlParsing /home/Paul/git/servo/android-toolchains/sdk-tools-linux-4333796/patcher/v4/package.xmlParsing /home/Paul/git/servo/android-toolchains/sdk-tools-linux-4333796/platform-tools/package.xmlParsing /home/Paul/git/servo/android-toolchains/sdk-tools-linux-4333796/platforms/android-25/package.xmlParsing /home/Paul/git/servo/android-toolchains/sdk-tools-linux-4333796/system-images/android-25/google_apis/armeabi-v7a/package.xmlParsing /home/Paul/git/servo/android-toolchains/sdk-tools-linux-4333796/tools/package.xmlAuto-selecting single ABI armeabi-v7a

Do you wish to create a custom hardware profile? [no]

```

Can we skip it somehow?

After bootstrap, the exports and the command line to start the emulator are printed. Would be great to explain what these do and what to do with it.

Also, can we make it so that we don't have to manually export them? When running `mach build --android`, we should check if `ANDROID_NDK/SDK` is set, and if not, check if `android-toolchains` is present?

**Comment by SimonSapin** [comment]
Made the custom hardware profile prompt automatic and silent (but not the license acceptance, so far). Change `bootstrap-android` to not print the environment variables, and made mach pick them up automatically when the relevant directories exist.

File: python/servo/bootstrap_commands.py [hunk 1]
```diff
@@ -54,6 +55,84 @@ def env(self):
...
+                image,
+            ])
+            subprocess.check_call([
+                path.join(tools_path, "tools", "bin", "avdmanager"),
+                "create", "avd",
+                "--path", path.join(toolchains, "avd", avd_name),
+                "--name", avd_name,
+                "--package", image,
+                "--force",
+            ])
```
[the remark this answers is not in the document]
Comment (SimonSapin) [reply python/servo/bootstrap_commands.py]: For using the host GPU, I think so. But that’s likely not the config we’ll use on CI.

File: python/servo/bootstrap_commands.py [hunk 2]
```diff
@@ -54,6 +55,84 @@ def env(self):
...
+    def bootstrap_android(self, update=False):
+
+        # Available filenames: see https://dl.google.com/android/repository/repository2-1.xml
+        ndk = "android-ndk-r12b-{system}-{arch}"
+        tools = "sdk-tools-{system}-4333796"
+
+        api_level = "25"
+        sdk_build_tools = "25.0.2"
+        system_image = "google_apis;armeabi-v7a"
+        avd_name = "servo-armv7"
```
[the remark this answers is not in the document]
Comment (SimonSapin) [reply python/servo/bootstrap_commands.py]: Done.

File: python/servo/package_commands.py [hunk 3]
```diff
@@ -416,7 +422,15 @@ def install(self, release=False, dev=False, android=False, target=None):
 
         if android:
             pkg_path = binary_path + ".apk"
-            exec_command = [self.android_adb_path(env), "install", "-r", pkg_path]
+            exec_command = [self.android_adb_path(env)]
+            if emulator and usb:
+                print("Cannot install to both emulator and USB at the same time.")
+                return 1
+            if emulator:
+                exec_command += ["-e"]
```
[the remark this answers is not in the document]
Comment (SimonSapin) [reply python/servo/package_commands.py]: Added `./mach android-emulator` which just forwards arguments to the `emulator` script. I’ll look later into having `./mach run` start it automatically.

[the remark this answers is not in the document]
Comment (SimonSapin) [reply python/servo/package_commands.py]: One difficulty is that after starting the emulator and after `adb wait-for-device` returns (which shows that `adb shell` is responsive), the userland might not have finished booting yet. In particular, there is an indeterminate amount of time until the Activity Manager is running and the `am start` command would success. This time can be especially long for emulator ARM. (Or x86 without working CPU acceleration.)

File: python/servo/bootstrap_commands.py [hunk 4]
```diff
@@ -54,6 +55,84 @@ def env(self):
...
+            if path.isdir(final):
+                return final
+
+            base_url = "https://dl.google.com/android/repository/"
+            filename = name + ".zip"
+            url = base_url + filename
+            archive = path.join(toolchains, filename)
+
+            if not path.isfile(archive):
+                download_file(filename, url, archive)
```
[the remark this answers is not in the document]
Comment (SimonSapin) [reply python/servo/bootstrap_commands.py]: Done.

File: support/android/egl-configs/.cargo/config [hunk 5]
```diff
@@ -0,0 +1,10 @@
+[build]
+target = "i686-linux-android"
+
+[target.armv7-linux-androideabi]
+linker = "./ld.sh"
+runner = "./run.sh"
+
+[target.i686-linux-android]
+linker = "./ld-i686.sh"
+runner = "./run.sh"
```
[the remark this answers is not in the document]
Comment (SimonSapin) [reply support/android/egl-configs/.cargo/config]: Per IRC discussion, I’ve removed all of `egl-config` now that it has served its purpose.

File: support/android/egl-configs/run.sh [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+# Specific device
+#target="-s something"
+
+# Emulator
+#target="-e"
+
+# USB
+#target="-d"
+
+path="${1}"
```
[the remark this answers is not in the document]
Comment (SimonSapin) [reply support/android/egl-configs/run.sh]: Per IRC discussion, I’ve removed all of `egl-config` now that it has served its purpose.

https://github.com/servo/servo/pull/21094#issuecomment-400097938 explains what this was about.

**Comment by SimonSapin** [comment]
```
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/buildslave/runprocess.py", line 426, in start
    self._startCommand()
  File "/usr/local/lib/python2.7/dist-packages/buildslave/runprocess.py", line 550, in _startCommand
    usePTY=self.usePTY)
  File "/usr/local/lib/python2.7/dist-packages/buildslave/runprocess.py", line 572, in _spawnProcess
    processProtocol, uid, gid, childFDs)
  File "/usr/local/lib/python2.7/dist-packages/twisted/internet/process.py", line 728, in __init__
    self._fork(path, uid, gid, executable, args, environment, fdmap=fdmap)
  File "/usr/local/lib/python2.7/dist-packages/twisted/internet/process.py", line 405, in _fork
    self.pid = os.fork()
OSError: [Errno 12] Cannot allocate memory
```

On the builder (`servo-linux-cross3`) the process `/usr/bin/python /usr/local/bin/buildslave start --nodaemon /home/servo/buildbot/slave` uses 4.7 GB of memory (out of less than 8 available total) with the CPU being idle (no job running). On `servo-linux-cross1` it’s 21 MB, for comparison.

CC @edunham, @aneeshusa

**Comment by asajeffrey** [comment]
@SimonSapin you said you got Servo to run in the android emulator with https://github.com/tomaka/glutin/pull/1036. It's merged into glutin 0.17.0, but compiling Servo with that version of glutin fails:
```
error[E0308]: mismatched types
   --> ports/servo/glutin_app/window.rs:214:47
    |
214 |             let glutin_window = GlWindow::new(window_builder, context_builder, &events_loop)
    |                                               ^^^^^^^^^^^^^^ expected struct `glutin::WindowBuilder`, found struct `winit::WindowBuilder`
    |
    = note: expected type `glutin::WindowBuilder`
               found type `winit::WindowBuilder`
note: Perhaps two different versions of crate `winit` are being used?
   --> ports/servo/glutin_app/window.rs:214:47
    |
214 |             let glutin_window = GlWindow::new(window_builder, context_builder, &events_loop)
    |                                               ^^^^^^^^^^^^^^

error[E0308]: mismatched types
   --> ports/servo/glutin_app/window.rs:214:80
    |
214 |             let glutin_window = GlWindow::new(window_builder, context_builder, &events_loop)
    |                                                                                ^^^^^^^^^^^^ expected struct `glutin::EventsLoop`, found struct `winit::EventsLoop`
    |
    = note: expected type `&glutin::EventsLoop`
               found type `&winit::EventsLoop`
note: Perhaps two different versions of crate `winit` are being used?
   --> ports/servo/glutin_app/window.rs:214:80
    |
214 |             let glutin_window = GlWindow::new(window_builder, context_builder, &events_loop)
    |                                                                                ^^^^^^^^^^^^

error[E0308]: mismatched types
   --> ports/servo/glutin_app/window.rs:224:17
    |
224 |             let (width, height) = glutin_window.get_inner_size().expect("Failed to get window inner size.");
    |                 ^^^^^^^^^^^^^^^ expected struct `glutin::dpi::LogicalSize`, found tuple
    |
    = note: expected type `glutin::dpi::LogicalSize`
               found type `(_, _)`

error[E0308]: mismatched types
   --> ports/servo/glutin_app/window.rs:297:21
    |
297 |                 let (_, height) = window.get_inner_size().expect("Failed to get window inner size.");
    |                     ^^^^^^^^^^^ expected struct `glutin::dpi::LogicalSize`, found tuple
    |
    = note: expected type `glutin::dpi::LogicalSize`
               found type `(_, _)`

error[E0061]: this function takes 1 parameter but 2 parameters were supplied
   --> ports/servo/glutin_app/window.rs:315:20
    |
315 |             window.set_inner_size(size.width as u32, size.height as u32)
    |                    ^^^^^^^^^^^^^^ expected 1 parameter

error[E0061]: this function takes 1 parameter but 2 parameters were supplied
   --> ports/servo/glutin_app/window.rs:322:20
    |
322 |             window.set_position(point.x as i32, point.y as i32)
    |                    ^^^^^^^^^^^^ expected 1 parameter

error[E0061]: this function takes 1 parameter but 2 parameters were supplied
   --> ports/servo/glutin_app/window.rs:540:28
    |
540 |                     window.resize(width, height);
    |                            ^^^^^^ expected 1 parameter

error[E0599]: no method named `hidpi_factor` found for type `&glutin::GlWindow` in the current scope
   --> ports/servo/glutin_app/window.rs:618:40
    |
618 |                 TypedScale::new(window.hidpi_factor())
    |                                        ^^^^^^^^^^^^
    |
    = help: did you mean `get_hidpi_factor`?

error[E0308]: mismatched types
   --> ports/servo/glutin_app/window.rs:677:35
    |
677 |                 window.set_cursor(winit_cursor);
    |                                   ^^^^^^^^^^^^ expected enum `glutin::MouseCursor`, found enum `winit::MouseCursor`
    |
    = note: expected type `glutin::MouseCursor`
               found type `winit::MouseCursor`
note: Perhaps two different versions of crate `winit` are being used?
   --> ports/servo/glutin_app/window.rs:677:35
    |
677 |                 window.set_cursor(winit_cursor);
    |                                   ^^^^^^^^^^^^

error[E0308]: mismatched types
   --> ports/servo/glutin_app/window.rs:694:21
    |
694 |                 let (width, height) = window.get_outer_size().expect("Failed to get window outer size.");
    |                     ^^^^^^^^^^^^^^^ expected struct `glutin::dpi::LogicalSize`, found tuple
    |
    = note: expected type `glutin::dpi::LogicalSize`
               found type `(_, _)`

error[E0308]: mismatched types
   --> ports/servo/glutin_app/window.rs:695:62
    |
695 |                 let (x, y) = window.get_position().unwrap_or((0, 0));
    |                                                              ^^^^^^ expected struct `glutin::dpi::LogicalPosition`, found tuple
    |
    = note: expected type `glutin::dpi::LogicalPosition`
               found type `({integer}, {integer})`

error[E0308]: mismatched types
   --> ports/servo/glutin_app/window.rs:695:21
    |
695 |                 let (x, y) = window.get_position().unwrap_or((0, 0));
    |                     ^^^^^^ expected struct `glutin::dpi::LogicalPosition`, found tuple
    |
    = note: expected type `glutin::dpi::LogicalPosition`
               found type `(_, _)`

error[E0308]: mismatched types
   --> ports/servo/glutin_app/window.rs:700:21
    |
700 |                 let (width, height) = window.get_inner_size().expect("Failed to get window inner size.");
    |                     ^^^^^^^^^^^^^^^ expected struct `glutin::dpi::LogicalSize`, found tuple
    |
    = note: expected type `glutin::dpi::LogicalSize`
               found type `(_, _)`
```
Did you cherry-pick that PR and apply it to glutin 0.13.1?

**Comment by SimonSapin** [comment]
@asajeffrey I first made that PR on top of 0.13.1 and tested that in Servo. I then rebased it on top of upstream/master in order to add the changelog entry.

The correct fix is to update glutin to 0.17, but to get something running in the meantime try cherry-picking and using `[patch]`. (Glutin upstream unfortunately stopped making git tags for releases, 0.13.1 is 1b2d62c0e9c78c340953aeea79cfe7d087e36aa7.)

**Comment by SimonSapin** [comment]
I ran this on `servo-linux-cross1.servo.org` (so hopefully without accidentally depending on stuff that happened to be installed on my machine):

```sh
git clone https://github.com/servo/servo
cd servo
git checkout android-bootstrap
git fetch origin pull/21102/head
git merge FETCH_HEAD --no-edit
```

(Get the repo with this PR + https://github.com/servo/servo/pull/21102)

```
export JAVA_HOME="/usr/lib/jvm/java-8-openjdk-amd64"
export PATH="$HOME/.cargo/bin:$PATH"
yes | ./mach bootstrap-android
./mach build --target i686-linux-android --dev
./mach package --target i686-linux-android --dev
./mach test-android-startup --dev
```

… and it failed at the last step with:

```
emulator: ERROR: x86 emulation currently requires hardware acceleration!
Please ensure KVM is properly installed and usable.
CPU acceleration status: KVM requires a CPU that supports vmx or svm
```

I assume that these CPU features are not available on AWS, or at least not in this machine.

On my Linux desktop this command ends with `JavaScript is running!`, which is a message that came from **JS running inside the Android emulator, through `console.log()` and `adb logcat`.

Once we figure out the CPU features on AWS, or switch to ARM and figure out GLES3 ¹, running `mach test-android-startup` on CI and gating PRs on it succeeding should catch the most egregious startup crashes.

(¹ GLES3 appears to require a system image for Android O or more recent, which is API level 26. But 25 seems to be the latest ARM image provided by Google.)

A next step will be running a subset of WPT, but the business of running the emulator and communicating with it will likely be similar.

**Comment by asajeffrey** [comment]
We could either try getting the emulator to work with software rendering, or run this on our mac CI which I believe has GPU hardware.

**Comment by SimonSapin** [comment]
I think that running on mac might work, but would be less desirable since the mac builder availability is already a resource constraint.

I’ll look into it some more, but so far it seems that even with software rendering GLES3 requires a more recent system image than is available for ARM. (https://stackoverflow.com/questions/43017539/cant-find-android-8-0-oreo-api-26-arm-system-images asks the same, https://developer.android.com/about/versions/oreo/android-8.0-migration mentions *Intel x86 Atom System Image* specifically.)

**Comment by asajeffrey** [comment]
Hopefully #21087 will land, and we'll get glutin 0.17.

**Review by paulrouget — COMMENTED** [review]
r=me with egl-configs mentions removed from git-ignore and Cargo.toml.

After this lands, can you update the wiki?

> Once we figure out the CPU features on AWS

Did you figure this out?

**Comment by francesca64** [comment]
@SimonSapin 
> Glutin upstream unfortunately stopped making git tags for releases

Fortunately, under my maintainership, that's no longer true: https://github.com/tomaka/glutin/tags

(Though that doesn't solve the issue of there being 3 years missing...)

**Comment by SimonSapin** [comment]
Thank you @francesca64 ! This will be helpful.

**Comment by SimonSapin** [comment]
@bors-servo r=paul

> Did you figure this out?

I’ve filed https://github.com/servo/servo/issues/21116 with some findings.

**Pull request merged**